[EVAL] Pointer value retrieval tasks

ggendro opened the pull request:
# Thank you for contributing an eval! ♥️

🚨 Please make sure your PR follows these guidelines, __failure to follow the guidelines below will result in the PR being closed automatically__. Note that even if the criteria are met, that does not guarantee the PR will be merged nor GPT-4 access granted. 🚨

__PLEASE READ THIS__:

In order for a PR to be merged, it must fail on GPT-4. We are aware that right now, users do not have access, so you will not be able to tell if the eval fails or not. Please run your eval with GPT-3.5-Turbo, but keep in mind as we run the eval, if GPT-4 gets higher than 90% on the eval, we will likely reject since GPT-4 is already capable of completing the task.

We plan to roll out a way for users submitting evals to see the eval performance on GPT-4 soon. Stay tuned! Until then, you will not be able to see the eval performance on GPT-4. **Starting April 10, the minimum eval count is 15 samples, we hope this makes it easier to create and contribute evals.**

Also, pelase note that we're using **Git LFS** for storing the JSON files, so please make sure that you move the JSON file to Git LFS before submitting a PR. Details on how to use Git LFS are available [here](https://git-lfs.com).

## Eval details 📑
### Eval name
pointer-value-retrieval

### Eval description

This benchmark evaluates the ability of a language model to extract a general pointer-value retrieval pattern from a set of few examples. The model must recognise the logical pattern in the examples and apply it to a test case. Each task consists in retrieving a value in an array according to a pointer to an index and a logic hidden to the model. We generate several splits. In the easy splits, the pointers and arrays are separated, while in the hard splits the pointers are hidden in the array. In the medium splits, the pointers are separated from the array, but the retrievals are made more complex. We also provide several versions varying the few-shot size.

### What makes this a useful eval?

This eval combines few-shot pattern recognition, abstract reasoning, and numerical reasoning; tasks on which large language models tend to perform poorly. This task requires precise reasoning that does not involve arithmetic or reading comprehension, which allows the evaluation to focus solely on the reasoning abilities of the model.

## Criteria for a good eval ✅

Below are some of the criteria we look for in a good eval. In general, we are seeking cases where the model does not do a good job despite being capable of generating a good response (note that there are some things large language models cannot do, so those would not make good evals).

Your eval should be:

- [X] Thematically consistent: The eval should be thematically consistent. We'd like to see a number of prompts all demonstrating some particular failure mode. For example, we can  create an eval on cases where the model fails to reason about the physical world.
- [X] Contains failures where a human can do the task, but either GPT-4 or GPT-3.5-Turbo could not.
- [X] Includes good signal around what is the right behavior. This means either a correct answer for `Basic` evals or the `Fact` Model-graded eval, or an exhaustive rubric for evaluating answers for the `Criteria` Model-graded eval.
- [X] **Include at least 15 high quality examples.**

If there is anything else that makes your eval worth including, please document it below.

### Unique eval value

> Insert what makes your eval high quality that was not mentioned above. (Not required)

## Eval structure 🏗️

Your eval should
- [X] Check that your data is in `evals/registry/data/{name}`
- [X] Check that your yaml is registered at `evals/registry/evals/{name}.yaml`
- [X] Ensure you have the right to use the data you submit via this eval

(For now, we will only be approving evals that use one of the existing eval classes. You may still write custom eval classes for your own cases, and we may consider merging them in the future.)

## Final checklist 👀

### Submission agreement

By contributing to Evals, you are agreeing to make your evaluation logic and data under the same MIT license as this repository. You must have adequate rights to upload any data used in an Eval. OpenAI reserves the right to use this data in future service improvements to our product. Contributions to OpenAI Evals will be subject to our usual Usage Policies (https://platform.openai.com/docs/usage-policies).

- [X] I agree that my submission will be made available under an MIT license and complies with OpenAI's usage policies.

### Email address validation

If your submission is accepted, we will be granting GPT-4 access to a limited number of contributors. Access will be given to the email address associated with the merged pull request.

- [X] I acknowledge that GPT-4 access will only be granted, if applicable, to the email address used for my merged pull request.

### Limited availability acknowledgement

We know that you might be excited to contribute to OpenAI's mission, help improve our models, and gain access to GPT-4. However, due to the requirements mentioned above and high volume of submissions, we will not be able to accept all submissions and thus not grant everyone who opens a PR GPT-4 access. We know this is disappointing, but we hope to set the right expectation before you open this PR.

- [X] I understand that opening a PR, even if it meets the requirements above, does not guarantee the PR will be merged nor GPT-4 access granted.

### Submit eval

- [X] I have filled out all required fields of this form
- [X] I have used **Git LFS** for the Eval JSON data
- [ ] (Ignore if not submitting code) I have run `pip install pre-commit; pre-commit install` and have verified that `black`, `isort`, and `autoflake` are running when I commit and push

Failure to fill out all required fields will result in the PR being closed.

### Eval JSON data 

Since we are using Git LFS, we are asking eval submitters to add in as many Eval Samples (at least 5) from their contribution here:

<details>
  <summary>View evals in JSON</summary>

  ### Eval
  ```jsonl
{"input": [{"role": "system", "content": "Figure out the pattern in the following examples and apply it to the test case. Your answer must follow the format of the examples."}, {"role": "system", "content": "(7,[2, 0, 8, 5, 3, 1, 3, 0, 7, 1]) -> 0"}, {"role": "system", "content": "(7,[4, 5, 2, 9, 3, 5, 0, 4, 8, 8]) -> 4"}, {"role": "system", "content": "(5,[8, 4, 3, 3, 4, 7, 8, 4, 8, 4]) -> 7"}, {"role": "system", "content": "(0,[3, 2, 4, 4, 2, 8, 7, 2, 3, 9]) -> 3"}, {"role": "system", "content": "(4,[7, 3, 4, 9, 5, 8, 0, 7, 9, 7]) -> "}], "ideal": 5}
{"input": [{"role": "system", "content": "Figure out the pattern in the following examples and apply it to the test case. Your answer must follow the format of the examples."}, {"role": "system", "content": "(1,[2, 3, 3, 2, 8, 3, 5, 9, 3, 8]) -> 3"}, {"role": "system", "content": "(9,[5, 8, 6, 5, 0, 1, 3, 8, 5, 6]) -> 6"}, {"role": "system", "content": "(1,[2, 0, 4, 8, 7, 3, 0, 1, 2, 0]) -> 0"}, {"role": "system", "content": "(1,[8, 2, 5, 1, 0, 6, 4, 4, 4, 5]) -> 2"}, {"role": "system", "content": "(3,[5, 7, 7, 3, 8, 5, 0, 4, 7, 8]) -> "}], "ideal": 3}
{"input": [{"role": "system", "content": "Figure out the pattern in the following examples and apply it to the test case. Your answer must follow the format of the examples."}, {"role": "system", "content": "(3,[8, 2, 7, 3, 6, 4, 5, 2, 2, 7]) -> 3"}, {"role": "system", "content": "(1,[8, 2, 0, 1, 7, 2, 3, 9, 0, 1]) -> 2"}, {"role": "system", "content": "(3,[8, 0, 3, 3, 8, 3, 2, 3, 6, 8]) -> 3"}, {"role": "system", "content": "(1,[6, 8, 4, 4, 5, 5, 5, 7, 1, 4]) -> 8"}, {"role": "system", "content": "(8,[5, 1, 1, 3, 9, 7, 3, 3, 7, 8]) -> "}], "ideal": 7}

{"input": [{"role": "system", "content": "Figure out the pattern in the following examples and apply it to the test case. Your answer must follow the format of the examples."}, {"role": "system", "content": "(6,[3, 9, 0, 6, 4, 1, 0, 8, 6, 8]) -> 8"}, {"role": "system", "content": "(2,[3, 5, 2, 9, 2, 0, 6, 9, 7, 9]) -> 9"}, {"role": "system", "content": "(3,[0, 4, 2, 8, 5, 9, 9, 0, 2, 4]) -> 9"}, {"role": "system", "content": "(2,[9, 8, 2, 1, 6, 1, 6, 7, 6, 8]) -> 6"}, {"role": "system", "content": "(5,[0, 9, 6, 3, 4, 4, 2, 0, 8, 4]) -> "}], "ideal": 4}
{"input": [{"role": "system", "content": "Figure out the pattern in the following examples and apply it to the test case. Your answer must follow the format of the examples."}, {"role": "system", "content": "(3,[1, 5, 3, 6, 7, 7, 1, 3, 3, 4]) -> 7"}, {"role": "system", "content": "(7,[8, 8, 9, 5, 2, 9, 8, 7, 4, 4]) -> 7"}, {"role": "system", "content": "(2,[2, 0, 3, 7, 4, 7, 7, 5, 7, 9]) -> 7"}, {"role": "system", "content": "(7,[1, 3, 9, 2, 6, 0, 3, 2, 2, 1]) -> 2"}, {"role": "system", "content": "(0,[8, 7, 6, 3, 2, 5, 6, 4, 6, 5]) -> "}], "ideal": 8}
{"input": [{"role": "system", "content": "Figure out the pattern in the following examples and apply it to the test case. Your answer must follow the format of the examples."}, {"role": "system", "content": "(7,[1, 0, 0, 0, 6, 1, 1, 0, 2, 8]) -> 8"}, {"role": "system", "content": "(0,[0, 0, 9, 3, 8, 5, 0, 6, 7, 6]) -> 9"}, {"role": "system", "content": "(3,[5, 3, 3, 3, 9, 9, 6, 9, 4, 5]) -> 9"}, {"role": "system", "content": "(2,[9, 5, 5, 0, 7, 3, 8, 2, 7, 9]) -> 7"}, {"role": "system", "content": "(7,[4, 0, 8, 1, 6, 6, 9, 8, 8, 9]) -> "}], "ideal": 9}

{"input": [{"role": "system", "content": "Figure out the pattern in the following examples and apply it to the test case. Your answer must follow the format of the examples."}, {"role": "system", "content": "[3, 7, 1, 4, 2, 5, 2, 4, 4, 5, 7] -> 2"}, {"role": "system", "content": "[7, 8, 3, 9, 8, 0, 9, 9, 0, 1, 9] -> 0"}, {"role": "system", "content": "[8, 1, 8, 4, 4, 4, 4, 9, 2, 9, 6] -> 9"}, {"role": "system", "content": "[0, 0, 9, 1, 5, 5, 2, 9, 6, 1, 2] -> 0"}, {"role": "system", "content": "[8, 8, 7, 8, 7, 1, 3, 7, 5, 9, 4] -> "}], "ideal": 9}
{"input": [{"role": "system", "content": "Figure out the pattern in the following examples and apply it to the test case. Your answer must follow the format of the examples."}, {"role": "system", "content": "[7, 3, 1, 5, 9, 1, 0, 9, 5, 1, 5] -> 5"}, {"role": "system", "content": "[8, 8, 7, 2, 2, 5, 1, 1, 8, 5, 0] -> 5"}, {"role": "system", "content": "[2, 9, 4, 0, 6, 9, 1, 7, 6, 4, 5] -> 0"}, {"role": "system", "content": "[9, 6, 4, 1, 6, 7, 9, 1, 7, 7, 8] -> 8"}, {"role": "system", "content": "[8, 2, 4, 5, 0, 1, 4, 4, 9, 6, 3] -> "}], "ideal": 6}
{"input": [{"role": "system", "content": "Figure out the pattern in the following examples and apply it to the test case. Your answer must follow the format of the examples."}, {"role": "system", "content": "[6, 5, 6, 2, 1, 1, 5, 5, 7, 8, 5] -> 5"}, {"role": "system", "content": "[0, 9, 7, 3, 8, 3, 6, 1, 6, 7, 5] -> 9"}, {"role": "system", "content": "[4, 7, 0, 0, 5, 6, 2, 8, 0, 7, 1] -> 6"}, {"role": "system", "content": "[3, 5, 4, 7, 9, 2, 1, 0, 7, 6, 7] -> 9"}, {"role": "system", "content": "[8, 9, 4, 8, 3, 8, 4, 1, 9, 4, 8] -> "}], "ideal": 4}
  ```
</details>

usama-openai reviewed (changes requested):
Thanks for the contribution. I would like to request the following changes:

1. I'm getting the following error while evaluating this PR.
    ```
    b'File "/content/evals/evals/eval.py", line 131, in eval_sample'
    b'return idx, self.eval_sample(sample, rng)'
    b'File "/content/evals/evals/elsuite/basic/match.py", line 45, in eval_sample'
    b'return evals.record_and_check_match('
    b'File "/content/evals/evals/api.py", line 84, in record_and_check_match'
    b'if not sampled.startswith(option):'
    b'TypeError: startswith first arg must be str or a tuple of str, not int'
    ```

    The error is occurring because the ideal answer should be added as a string, but in all the datasets, the ideal answers are added as an integer.

2. I would recommend that you create an eval set for these evals. You can read documentation or have a look at other eval sets in the `evals/evals/registry/eval_sets/` directory.

3. Add a proper description for each eval in the `.yaml` file.

We would love to review the PR again after the suggested changes.

ggendro commented:
Thank you for the feedback!

I proceeded to the following changes:
 1. I changed the format of the ideal labels from int to str
 2. I added a file in evals_sets listing the new evals, following the format of the other eval sets
 3. I added a description for each eval file

usama-openai reviewed (changes requested):
Thanks for implementing the requested changes.

I would recommend asking the model to reason before answering and using the `Includes` evaluation method here because such complex tasks are hard for the model to do zero-shot without a chance to reason through the steps. Asking the model to provide reasoning will give the model a fair chance to solve the puzzle, and a proper evaluation method can be used to evaluate the provided answer. You also need to add instructions to provide output in a specific format, and then use that format to write the ideal answer. For example, the model can be asked to enclose the final answer in square brackets, and the ideal answer can be formatted like `[5]`. It'll give the model a proper chance to reason before answering, and the proper formatting will help in identifying the final answer properly using the `Includes` method.

ggendro commented:
Thank you for the additional feedback. I modified the evaluation from `Match` to `Includes` and changed the prompt to `Figure out the pattern in the following examples and apply it to the test case. Explain your reasoning step-by-step and provide the expected output immediately after writing the word ANSWER: .`.
I think this method make it easier to extract the actual answer. I found that ChatGPT tends to attempt to explain each of the examples, so it's hard to disambiguate the actual answer for the test case from the answer of a previous example. Asking to write the answer after the keyword `ANSWER: ` solves this issue.

usama-openai reviewed (approved):
This PR looks in good shape now. I'm approving this PR.

usama-openai commented:
You should see GPT-4 API access enabled in your account in the next few days.
